lib/SPIRV/SPIRVReader.cpp

```diff
@@ -816,6 +816,36 @@ Value *SPIRVToLLVM::transConvertInst(SPIRVValue *BV, Function *F,
   return ConstantExpr::getCast(CO, dyn_cast<Constant>(Src), Dst);
 }
 
+static void applyNoIntegerWrapDecorations(const SPIRVValue *BV,
+                                          Instruction *Inst) {
+  if (BV->hasDecorate(DecorationNoSignedWrap)) {
+    Inst->setHasNoSignedWrap(true);
+  }
+
+  if (BV->hasDecorate(DecorationNoUnsignedWrap)) {
+    Inst->setHasNoUnsignedWrap(true);
+  }
+}
+
+static void applyFPFastMathModeDecorations(const SPIRVValue *BV,
+                                           Instruction *Inst) {
+  SPIRVWord V;
+  FastMathFlags FMF;
+  if (BV->hasDecorate(DecorationFPFastMathMode, 0, &V)) {
+    if (V & FPFastMathModeNotNaNMask)
+      FMF.setNoNaNs();
+    if (V & FPFastMathModeNotInfMask)
+      FMF.setNoInfs();
+    if (V & FPFastMathModeNSZMask)
+      FMF.setNoSignedZeros();
+    if (V & FPFastMathModeAllowRecipMask)
+      FMF.setAllowReciprocal();
+    if (V & FPFastMathModeFastMask)
```

Comment:
Can someone please explain why `SPIR-V` `Fast` implies `LLVM` `fast`?  `SPIR-V` `Fast` [means](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#FP_Fast_Math_Mode) `NotNaN|NotInf|NSZ|AllowRecip`, while LLVM 'fast' [means](https://llvm.org/docs/LangRef.html#fast-math-flags) `nnan|ninf|nsz|arcp|`**contract|afn|reassoc**.

Reply:
Tagging @StuartDBrady here for visibility

Reply:
> I would assume that somewhere in SPIR-V Writer we would map LLVM `fast` into SPIR-V `fast`, so, in order to preserve LLVM IR -> SPIR-V -> LLVM IR roundtrip we use this particular mapping here.

Yes, the round-trip mapping is the reason.

For SPIR-V which _specifically_ requires the `NotNaN`, `NotInf`, `NSZ` and `AllowRecip` flags and no others, `NotNaN|NotInf|NSZ|AllowRecip` should be used. The entire point of having a `Fast` flag is to allow extra flags to be added to it.

Reply:
Then I am concerned about the documentation of `SPIR-V` `Fast`, which says:

> Allow algebraic transformations according to real-number associative and distributive algebra. This flag implies all the others.

Are you saying that `all the others` means `any possible` FP optimization flags existing in the particular `SPIR-V` consumer implementation?  It does not sound very robust for the spec.  I also do not see any comments explaining that `SPIR-V` producers complying with the spec must not use `Fast` as a replacement of `NotNaN|NotInf|NSZ|AllowRecip`.  Did I miss it?

Reply:
My understading is:
`all the others` == `NotNaN|NotInf|NSZ|AllowRecip` 
`associative and distributive algebra` == `contract|afn|reassoc`

Reply:
I agree that "associative ... algebra" includes `reassoc`, but I think `contract` and `afn` are not covered by this.  What is the right way to get an explanation of the SPIR-V specification, if I have doubts?

Reply:
>  What is the right way to get an explanation of the SPIR-V specification, if I have doubts?

You could ask for the clarification: I'm not aware of the open-source repo or bug tracker related to core SPIR-V spec, but you can try creating issues in https://github.com/KhronosGroup/SPIRV-Headers or https://github.com/KhronosGroup/SPIRV-Registry. And for sure the best way is probably to discuss this in internal KhronosGroup repos corresponding to SPIR-V spec